src/main/java/net/minestom/server/network/packet/server/play/ParticlePacket.java

```diff
@@ -3,21 +3,39 @@
 import net.minestom.server.network.NetworkBuffer;
 import net.minestom.server.network.packet.server.ServerPacket;
 import net.minestom.server.network.packet.server.ServerPacketIdentifier;
+import net.minestom.server.particle.ParticleOption;
+import net.minestom.server.utils.binary.BinaryWriter;
 import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Objects;
 
 import static net.minestom.server.network.NetworkBuffer.*;
 
-public record ParticlePacket(int particleId, boolean longDistance,
-                             double x, double y, double z,
+public record ParticlePacket(int particleId, boolean longDistance, double x, double y, double z,
                              float offsetX, float offsetY, float offsetZ,
-                             float particleData, int particleCount, byte[] data) implements ServerPacket {
+                             float particleData, int particleCount,
+                             byte[] data) implements ServerPacket {
+
     public ParticlePacket(@NotNull NetworkBuffer reader) {
         this(reader.read(VAR_INT), reader.read(BOOLEAN),
                 reader.read(DOUBLE), reader.read(DOUBLE), reader.read(DOUBLE),
                 reader.read(FLOAT), reader.read(FLOAT), reader.read(FLOAT),
                 reader.read(FLOAT), reader.read(INT), reader.read(RAW_BYTES));
     }
 
+    public ParticlePacket(int particleId, boolean longDistance, double x, double y, double z,
```

Comment:
I would recommend against having ParticleOption inside ParticlePacket.
ParticleOption should be part of the ParticleCreator API, not the network protocol.

I'd prefer it if the original packet definition is used (with byte[] data) and the byte[] data is calculated in ParticleCreator#createParticlePacket

Reply:
Done